docs/programmers_guide/witness_formal_spec.md

```diff
@@ -0,0 +1,196 @@
+# Block Witness Formal Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## The Stack
+
+Witnesses are executed on a stack machine that builds tries/calculates hashes.
+
+The stack consists of pairs `(node, hash)`.
+
+Each witness is a queue of instructions.
+In every execution cycle one instruction gets dequeued and a matching substitution rule gets applied to the stack.
+
+In the end, when there are no more instructions left, there should be only one item on the stack.  
+
+## Substitution rules
+
+Here is an example substitution rule. The current instruction is named `INSTR1`.
+
+```
+STACK(node1, hash1) STACK(node2, hash2) INSTR1(params) |=> 
+STACK(helper1(node1, node2), helper2(hash1, hash2))
+```
+
+This substitution rule replaces 2 stack items `(node1, hash1)` and `(node2, hash2)`
+with a single stack item `(helper1(node1, node2), helper2(hash1, hash2))`.
+
+Where `helper1` and `helper2` are pure functions defined in pseudocode (see the example below).
+
+```
+helper1 (value1, value2) {
+    return value1 + value2
+}
+
+helper2 (hash1, hash2) {
+    return KECCAK(hash1 + hash2)
+}
+```
+
+## Guards & Traps
+
+Each instruction and each helper function can have zero, one or multiple `guard` statements.
+Each `guard` statement looks like this:
+
+```
+guard <CONDITION> else {
+    TRAP
+}
+```
+
+That means that for this instruction to be valid the `<CONDITION>` in the guard statement must be true.
+
+If the condition is falsey, then the execution should stop (`TRAP` instruction). Implementation of the `TRAP` instruction is undefined, but it should stop the execution flow. (e.g. in Golang it might be a function returning an error or a panic).
+
+If an instruction has multiple guard statements, all of the conditions specified there should be satisfied.
+
+## Instructions & Parameters
+
+Each instruction can have zero, one or multiple parameters. The values of these parameters are stored in the witness itself and decoded together with the instruction. They aren't taken from the stack.
+
+That makes it different from the helper function parameters that can come from either stack or the witness itself.
+
+## Helper functions
+
+Helper functions are pure functions that are used in the substitution rules.
+They can have zero, one or multiple arguments.
```

Comment:
It does not make sense for a helper function to have zero arguments, because if it is pure function, it is a constant :)

Reply:
you are right :)